### PR TITLE
pcsx2-gui: Internalize VirtualPad and Upgrade NewRecordingFrame

### PR DESCRIPTION
**Pull request by sonicfind**
Simplifies virtualPad access by solely holding the virtualPad window pointers in the g_inputRecording object. This will be useful later down the line when more virtualPads are added.

NewRecordingFrame has improved functionality:
* Users can now only choose to create a savestate-based recording if GS is open.
* Forced file browsing in certain situations.

Also, forces a game to have been loaded before being allowed to mess with input recordings.

### CODE REVIEW

File: pcsx2/Recording/InputRecording.cpp [hunk 1]
```diff
@@ -63,28 +63,51 @@ void SaveStateBase::InputRecordingFreeze()
...
 
-InputRecording::InputRecording()
+InputRecording::~InputRecording()
 {
-	// NOTE - No multi-tap support, only two controllers
-	padData[CONTROLLER_PORT_ONE] = new PadData();
-	padData[CONTROLLER_PORT_TWO] = new PadData();
+	// As to not attempt a double delete from WX
+	newRecordingFrame.release();
+	openFileDialog.release();
```
Comment (xTVaser) [inline pcsx2/Recording/InputRecording.cpp:70]: What's the scenario that leads to this double delete you speak of?  Once the object is destroyed these smart pointers should get automatically cleaned up should they not?

File: pcsx2/Recording/InputRecording.cpp [hunk 2]
```diff
@@ -63,28 +63,51 @@ void SaveStateBase::InputRecordingFreeze()
...
 {
-	virtualPads[port] = ptr;
+	padData = std::make_unique<PadData>();
 }
 
-void InputRecording::RecordingReset()
+InputRecording::InputRecordingPad::~InputRecordingPad()
+{
+	// As to not attempt a double delete from WX
+	virtualPad.release();
```
Comment (xTVaser) [inline pcsx2/Recording/InputRecording.cpp:81]: same question here

File: pcsx2/Recording/InputRecording.cpp [hunk 3]
```diff
@@ -63,28 +63,51 @@ void SaveStateBase::InputRecordingFreeze()
...
+}
+
+void InputRecording::InitInputRecordingWindows(wxWindow* parent)
+{
+	if (newRecordingFrame.get() != nullptr)
+		return;
```
Comment (xTVaser) [inline pcsx2/Recording/InputRecording.cpp:87]: this fundamentally works, but seems like a bit of an anti-pattern = using the initialization of 1/4 of the preceding variables as a check for all of them.  

I'd much rather see checking before initializing each, or have an all-encompassing `bool` to check if Recording GUI elements have already been initialized.

File: pcsx2/Recording/InputRecording.cpp [hunk 4]
```diff
@@ -101,7 +124,7 @@ void InputRecording::ControllerInterrupt(u8& data, u8& port, u16& bufCount, u8 b
 		fInterruptFrame = false;
 
 	// We do not want to record or save the first two bytes in the data returned from the PAD plugin
-	else if (fInterruptFrame && bufCount >= 3 && frameCounter >= 0 && frameCounter < INT_MAX)
+	else if (fInterruptFrame && frameCounter >= 0 && frameCounter < INT_MAX)
```
Comment (xTVaser) [inline pcsx2/Recording/InputRecording.cpp:127]: this change appears unrelated in the scope of this PR

File: pcsx2/Recording/InputRecording.cpp [hunk 5]
```diff
@@ -220,28 +239,32 @@ wxString InputRecording::RecordingModeTitleSegment()
...
+	pads[CONTROLLER_PORT_ONE].virtualPad->SetReadOnlyMode(true);
+	pads[CONTROLLER_PORT_TWO].virtualPad->SetReadOnlyMode(true);
 	inputRec::log("Replay mode ON");
 }
 
 void InputRecording::SetFrameCounter(u32 newGFrameCount)
 {
-	if (newGFrameCount > startingFrame + (u32)g_InputRecording.GetInputRecordingData().GetTotalFrames())
+	const u32 endFrame = startingFrame + inputRecordingData.GetTotalFrames();
+	if (newGFrameCount >= endFrame)
```
Comment (xTVaser) [inline pcsx2/Recording/InputRecording.cpp:258]: hmm, why the change to `>=` does `GetTotalFrames` return something different now? (havn't gotten to that point in the review yet) or does this resolve a bug of some sort?

File: pcsx2/Recording/InputRecording.cpp [hunk 6]
```diff
@@ -298,8 +324,11 @@ bool InputRecording::Create(wxString FileName, bool fromSaveState, wxString auth
 	inputRecordingData.GetHeader().SetEmulatorVersion();
 
 	// Set author name
-	if (!authorName.IsEmpty())
-		inputRecordingData.GetHeader().SetAuthor(authorName);
+	{
+		const wxString authorName = newRecordingFrame->GetAuthor();
+		if (!authorName.IsEmpty())
+			inputRecordingData.GetHeader().SetAuthor(authorName);
+	}
```
Comment (xTVaser) [inline pcsx2/Recording/InputRecording.cpp:331]: these brace brackets appear to be wrapping nothing

File: pcsx2/Recording/InputRecording.cpp [hunk 7]
```diff
@@ -308,17 +337,20 @@ bool InputRecording::Create(wxString FileName, bool fromSaveState, wxString auth
 	SetToRecordMode();
 	g_InputRecordingControls.DisableFrameAdvance();
 	inputRec::log("Started new input recording");
-	inputRec::consoleLog(fmt::format("Filename {}", std::string(FileName)));
+	inputRec::consoleLog(fmt::format("Filename {}", std::string(inputRecordingData.GetFilename())));
 	return true;
 }
 
-bool InputRecording::Play(wxString fileName)
+int InputRecording::Play()
```
Comment (xTVaser) [inline pcsx2/Recording/InputRecording.cpp:344]: If we need to return more than 2 states, I'd rather it return an `enum class` so that is reads better on the consuming side.  Returning hard-coded `int`s requires knowing the internals of this function.  Though maybe it's an indication of this function trying to do too much on it's own

File: pcsx2/Recording/InputRecording.cpp [hunk 8]
```diff
@@ -361,6 +393,27 @@ bool InputRecording::Play(wxString fileName)
...
+}
+
+bool InputRecording::GoToFirstFrame()
+{
+	if (inputRecordingData.FromSaveState())
+	{
+		if (!wxFileExists(inputRecordingData.GetFilename() + "_SaveState.p2s"))
+		{
+			recordingConLog(wxString::Format("[REC]: Could not locate savestate file at location - %s_SaveState.p2s\n",
+												inputRecordingData.GetFilename()));
```
Comment (xTVaser) [inline pcsx2/Recording/InputRecording.cpp:406]: use the new `inputRec::consoleLog`

File: pcsx2/gui/AppInit.cpp [hunk 9]
```diff
@@ -77,19 +72,6 @@ void Pcsx2App::OpenMainFrame()
 	DisassemblyDialog* disassembly = new DisassemblyDialog( mainFrame );
 	m_id_Disassembler = disassembly->GetId();
 
-#ifndef DISABLE_RECORDING
```
Comment (xTVaser) [inline pcsx2/gui/AppInit.cpp:80]: Seems like the new location for initializing the recording windows is now a side-effect of the menubar being created?  Not sure how I feel about that, that feels unexpected and should probably remain here where other frames (like the debugger/console) are made.

File: pcsx2/gui/FrameForGS.cpp [hunk 10]
```diff
@@ -134,18 +136,23 @@ void GSPanel::InitRecordingAccelerators()
...
+		GetAssociatedKeyCode("InputRecordingModeToggle"),
 		g_InputRecording.IsActive());
 
 	inputRec::consoleLog("Initialized Input Recording Key Bindings");
 }
 
+wxString GSPanel::GetAssociatedKeyCode(const char* id)
+{
+	return m_Accels->findKeycodeWithCommandId(id).toTitleizedString();
+}
```
Comment (xTVaser) [inline pcsx2/gui/FrameForGS.cpp:154]: 👍 though I wonder if this could go into the `AcceleratorDictionary` class and be used generically outside of this file.  Also `GetAssociatedKeyCode` is a bit misleading as it's already converting the KeyCode into it's string representation.  This is mostly a nit, feel free to ignore.

File: pcsx2/gui/GlobalCommands.cpp [hunk 11]
```diff
@@ -531,6 +538,15 @@ namespace Implementations
...
 	}
 
+	void GoToFirstFrame()
+	{
+		if (g_Conf->EmuOptions.EnableRecordingTools && g_InputRecording.IsActive())
+		{
+			if (!g_InputRecording.GoToFirstFrame() && wxGetApp().HasGUI())
+				sMainFrame.RecordingMenuReset();
+		}
+	}
```
Comment (xTVaser) [inline pcsx2/gui/GlobalCommands.cpp:548]: needs to be wrapped in `ifndef`

File: pcsx2/gui/MainFrame.cpp [hunk 12]
```diff
@@ -57,11 +61,18 @@ void MainEmuFrame::UpdateStatusBar()
 {
 	wxString temp(wxEmptyString);
 
-	if (g_Conf->EnableFastBoot)
-		temp += "Fast Boot - ";
+#ifndef DISABLE_RECORDING
+	if (g_InputRecording.IsActive() && g_InputRecording.GetInputRecordingData().FromSaveState())
+		temp += "Base Savestate - " + g_InputRecording.GetInputRecordingData().GetFilename() + "_SaveState.p2s";
+	else
```
Comment (xTVaser) [inline pcsx2/gui/MainFrame.cpp:67]: I'm not sure how I feel about modifying the system menu.  I mean it makes sense for what you're trying to accomplish.  But I can't help but feel like a better way to go would be to:
- gray out the boot option / disable it
- add a base savestate option to the `Input Record` menu.
To me it feels better UX-wise, but I could very well be wrong.

File: pcsx2/gui/MainFrame.cpp [hunk 13]
```diff
@@ -57,11 +61,18 @@ void MainEmuFrame::UpdateStatusBar()
...
+#endif
+	{
+		if (g_Conf->EnableFastBoot)
+			temp += "Fast Boot - ";
 
-	if (g_Conf->CdvdSource == CDVD_SourceType::Iso)
-		temp += "Load: '" + wxFileName(g_Conf->CurrentIso).GetFullName() + "' ";
+		if (g_Conf->CdvdSource == CDVD_SourceType::Iso)
+			temp += "Load: '" + wxFileName(g_Conf->CurrentIso).GetFullName() + "' ";
+	}
```
Comment (xTVaser) [inline pcsx2/gui/MainFrame.cpp:75]: i think these brackets will cause an issue if the `#ifndef` isn't present.  We'll have dangling brackets without an associated else...maybe that's valid considering my previous comment but it reads strangely currently.

File: pcsx2/gui/MainFrame.cpp [hunk 14]
```diff
@@ -92,6 +103,12 @@ void MainEmuFrame::UpdateCdvdSrcSelection()
 			jNO_DEFAULT
 	}
 	sMenuBar.Check(cdsrc, true);
+#ifndef DISABLE_RECORDING
+	if (g_InputRecording.IsActive())
+		ApplyFirstFrameStatus();
+	else
+#endif
+		ApplyCDVDStatus();
```
Comment (xTVaser) [inline pcsx2/gui/MainFrame.cpp:111]: indentation here seems to be messed up

File: pcsx2/gui/MainFrame.cpp [hunk 15]
```diff
@@ -483,16 +500,18 @@ void MainEmuFrame::CreateCaptureMenu()
...
+	m_menuRecording.Append(MenuId_Recording_Play, _("Play"), _("Playback an existing input recording."))->Enable(false);
 	m_menuRecording.AppendSeparator();
 	m_menuRecording.Append(MenuId_Recording_TogglePause, _("Toggle Pause"), _("Pause or resume emulation on the fly."))->Enable(false);
 	m_menuRecording.Append(MenuId_Recording_FrameAdvance, _("Frame Advance"), _("Advance emulation forward by a single frame at a time."))->Enable(false);
 	m_menuRecording.Append(MenuId_Recording_ToggleRecordingMode, _("Toggle Recording Mode"), _("Save/playback inputs to/from the recording file."))->Enable(false);
 	m_menuRecording.AppendSeparator();
 	m_menuRecording.Append(MenuId_Recording_VirtualPad_Port0, _("Virtual Pad (Port 1)"));
 	m_menuRecording.Append(MenuId_Recording_VirtualPad_Port1, _("Virtual Pad (Port 2)"));
+	if (g_Conf->EmuOptions.EnableRecordingTools)
+		g_InputRecording.InitInputRecordingWindows(this);
```
Comment (xTVaser) [inline pcsx2/gui/MainFrame.cpp:514]: once again, I'm not sure if this is where we want to do this.  This function is supposed to be for initializing the wxMenuBar, I don't think we want to overload it with other tasks

File: pcsx2/gui/MainFrame.cpp [hunk 16]
```diff
@@ -825,9 +851,9 @@ void MainEmuFrame::initializeRecordingMenuItem(MenuIdentifiers menuId, wxString
 	item.Enable(enable);
 }
 
-void MainEmuFrame::enableRecordingMenuItem(MenuIdentifiers menuId, bool enable)
+void MainEmuFrame::enableRecordingMenuItem(int menuId, bool enable)
 {
-	wxMenuItem& item = *m_menuRecording.FindChildItem(menuId);
+	wxMenuItem& item = *m_menuRecording.FindItem(menuId);
```
Comment (xTVaser) [inline pcsx2/gui/MainFrame.cpp:856]: can i get some context on these changes.  From my perspective using the `enum` added some extra type-safety, and `FindChildItem` makes sense as it's specifically for finding menu sub-menu elements, is this to support finding nested menu elements in the future?

File: pcsx2/gui/MainFrame.h [hunk 17]
```diff
@@ -255,6 +261,7 @@ class MainEmuFrame : public wxFrame,
 	void Menu_Recording_New_Click(wxCommandEvent& event);
 	void Menu_Recording_Play_Click(wxCommandEvent& event);
 	void Menu_Recording_Stop_Click(wxCommandEvent& event);
+	void ApplyFirstFrameStatus();
```
Comment (xTVaser) [inline pcsx2/gui/MainFrame.h:264]: does this need to be wrapped in an `ifndef`?

File: pcsx2/gui/MainMenuClicks.cpp [hunk 18]
```diff
@@ -945,65 +1019,93 @@ void MainEmuFrame::Menu_Recording_New_Click(wxCommandEvent& event)
...
-			g_InputRecordingControls.Resume();
-		return;
-	}
 
-	wxString path = openFileDialog.GetPath();
 	const bool recordingLoaded = g_InputRecording.IsActive();
-	if (!g_InputRecording.Play(path))
+	switch (g_InputRecording.Play())
 	{
+	case 1:
```
Comment (xTVaser) [inline pcsx2/gui/MainMenuClicks.cpp:1043]: rather switching on enum values as mentioned above

File: pcsx2/gui/MainMenuClicks.cpp [hunk 19]
```diff
@@ -945,65 +1019,93 @@ void MainEmuFrame::Menu_Recording_New_Click(wxCommandEvent& event)
...
 	{
 		m_menuRecording.FindChildItem(MenuId_Recording_New)->Enable(false);
 		m_menuRecording.FindChildItem(MenuId_Recording_Stop)->Enable(true);
+		m_menuRecording.FindChildItem(MenuId_Recording_ToggleRecordingMode)->Enable(true);
 	}
-	sMainFrame.enableRecordingMenuItem(MenuId_Recording_ToggleRecordingMode, g_InputRecording.IsActive());
+	ApplyFirstFrameStatus();
+}
+
+void MainEmuFrame::ApplyFirstFrameStatus()
```
Comment (xTVaser) [inline pcsx2/gui/MainMenuClicks.cpp:1061]: wrapped in an ifndef?

**Review by xTVaser — COMMENTED** [review]
Overall this looks good, lots of nice improvements. With that said, I don't mind reviewing/testing large PRs like this but it would make things a lot simpler and faster if we could reduce the surface area per-PR. ie, this PR changes several things that could be broken apart:
- reducing core footprint
- adding new GUI options
- looks like a handful of bug fixes / alterations inside the recording logic itself, etc.

File: pcsx2/Recording/InputRecording.cpp (hunk 1 above)
Comment (sonicfind) [reply pcsx2/Recording/InputRecording.cpp:70]: When the new operator is called on a wxWindow object, the point of the new object gets sent into a list of that all wxWindow pointers. WX deletes every pointer in that list, so if the smart pointer deletes it first...

File: pcsx2/gui/MainFrame.cpp (hunk 15 above)
Comment (sonicfind) [reply pcsx2/gui/MainFrame.cpp:514]: Fair.

File: pcsx2/Recording/InputRecording.cpp (hunk 7 above)
Comment (sonicfind) [reply pcsx2/Recording/InputRecording.cpp:344]: Well luckily, it's getting reworked back into a bool thanks to changes to MainFrame::SetRecordingMenu (or RecordingMenuReset in the previous iteration).

**Review by xTVaser — APPROVED** [review]
Tested the following:
- VirtualPad still opens and functions correctly with new adjustments
- Browsing for the recording's file-path was required, even manually deleting the field properly updated the UI
- Was only able to make a power-on or "boot" recording with GS closed
- New/Play options were disabled until i booted a game for the first time upon opening the emulator.

Looks good to me.
